### DISCUSSION TITLE
More graphics API support

### DISCUSSION THREAD

**Discussion by L2750558108** [godotengine/godot-proposals#6045] [Engine Core]
Support more graphics APIs like Unity does and allow automatic selection (selection that works and provides optimal performance on the current platform) or sequential fallback (the user provides a list of graphics APIs, giving preference to the top graphics APIs, and then selecting down if not) for better performance and compatibility

![image](https://user-images.githubusercontent.com/119792950/210734946-1c5f873c-bf01-4c4f-bae4-8c16f64f7fd2.png) ![image](https://user-images.githubusercontent.com/119792950/210734956-2f59303c-0997-45f3-8c9a-2bdaf74456b6.png)

**Comment by Zireael07** [comment]
Supporting Direct3D is a MASSIVE undertaking.

Godot already supports Vulkan and GLES3.

**Reply by L2750558108** [reply]
Is there a plan for it? And what about automatically selecting the best performant renderer and fallback list?

**Reply by AThousandShips** [reply]
There's a PR open for [Direct3D](https://github.com/godotengine/godot/pull/70315)

**Reply by L2750558108** [reply]
So what about automatically selecting the best performant renderer and fallback list?

**Reply by AThousandShips** [reply]
I don't know how difficult detecting the appropriate setup for a particular computer, but there would of course be management of what renderer is available and such.

**Reply by Calinou** [reply]
> So what about automatically selecting the best performant renderer and fallback list?

In Godot, Vulkan is expected to remain faster than Direct3D 12 for the foreseeable future. One exception might be Intel IGPs on Windows (where Vulkan exhibits performance issues most often), but it needs to be thoroughly tested.

**Reply by L2750558108** [reply]
These two features prevent compatibility issues. The Auto Select Best Graphics API feature will run the available graphics APIs that do not support Vulkan devices to automatically select the best performing graphics. The fallback list feature will allow developers to customize which graphics API to fall back to when the graphics API is not supported.
